### PR TITLE
Add --blob-exec to run system commands for each blob

### PR DESCRIPTION
**Pull request by nightscape**
This is a rebase of Paul Draper's implementation of blob-exec:
https://github.com/rtyley/bfg-repo-cleaner/pull/83

Is there any chance to merge this? I'm currently using it to replace tabs by spaces exactly like Paul mentioned it under his use cases.

### CODE REVIEW

**Comment by javabrett** [comment]
I'm interested in whether there are some metrics that show how much faster this is than running the equivalent filter-branch, if anyone had those numbers.

**Comment by OwnageIsMagic** [comment]
It doesnt work for me
image
What I am missing?

**Comment by dmgerman** [comment]
> I'm interested in whether there are some metrics that show how much faster this is than running the 
> equivalent filter-branch, if anyone had those numbers.

given my experience with this branch (running on linux) and filter-branch, my view is that the savings are a function that depends:

1. On the number of commits C
2. On the number of unique files in the project UF
3. On the number of files replaced in a given commit 
4. On the number of files that exist in the file system at any given time
6. the speed of the file system,
7. the time it takes to process each individual file.

With BFG, the cost of the replacement is a function of (k1 * C + k2 * UF)/(k3*SpeedFileSystem)

With filter-branch, this is what happens:

For every commit:
   - checkout and replace files that are different in the file system with respect to their contents according to the commit
   - run filter branch on each file (no duplicates detected)
   - create new commit

For example, say a filter replaces the contents of every file in every revision.

For revision 1 the contents are checked-out. Lets say we have n1 files.
n1 files are replaced with new versions.
Commit these changes.
Now, this is the true expensive operation: checkout the next commit. This means, replacing every single file in the tree (even those that were not part of the actual changes) with their version according to the commit. Now we have to process _again_ every single file in the repo (they are all dirty). After they are all processed, git will realize that only the files that were in commit 2 are actually changed (with respect to the processed version of commit 1)

This process basically makes filter-branch cost include:
- cost of recreating every file in every revision (cost of checking out and writing files in repo--avg files in repo * # of commits)
- cost of processing every file in every revision (cost of processing: avg files in repo * # commits)

so, bfg is proportional to the avg number of files in commit multiplied by the number of commits (basically, the number of unique BLOB files found in a repo) while the cost of filter-branch is proportional to th avg number of files _IN_ repo multiplied by number of commits.

Let's say we can do 100 file processed operations per second. And we have 1M commits, with an average of 10k files in the repo, and 10 files modified per commit (I am thinking linux here).

Let us assume that the cost of checking out the files (filter-branch) and processing the commits (bfg and filter-branch) is neglegible (it is not, but bear with me). if my numbers are right, it would take:

(1M * 10k)/100 seconds to process this repo => 1157 days => 3 years

with BFG it would take:

(1M * 10) /100 seconds => 27 hrs.

so, in conclusion: BFG processing time is Order(#commits *#filesPerCommit)
while the filter-branch processing time is Order(#commits * #avgNumberOfFilesInRepo).
